### PR TITLE
Move metric definitions to mlflow.metrics

### PR DESCRIPTION
**Pull request by annzhang-db**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

- use make_metric for existing text metrics
- add documentation for said metrics

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [x] API references
  - [ ] Instructions

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/metrics/base.py [hunk 1]
```diff
@@ -14,3 +21,306 @@ class MetricValue:
...
+
+
+# general text metrics
+toxicity = make_metric(
+    eval_fn=_toxicity_eval_fn,
+    greater_is_better=False,
+    name="toxicity",
+    long_name="toxicity/roberta-hate-speech-dynabench-r4",
+    version="v1",
+)
```
Comment (dbczumar) [inline mlflow/metrics/base.py:269]: Can we add docstrings for each of these?

File: mlflow/metrics/base.py [hunk 2]
```diff
@@ -14,3 +21,306 @@ class MetricValue:
     scores: List[float] = None
     justifications: List[float] = None
     aggregate_results: Dict[str, float] = None
+
+
+def _validate_text_predictions(predictions, metric_name):
```
Comment (sunishsheth2009) [inline mlflow/metrics/base.py:26]: Can we move all of these metrics into a new file or else this file would be huge. It also be utils where we can import this.

In future, we can add other precanned metrics a similar manner as well

File: mlflow/metrics/base.py (hunk 1 above)
Comment (annzhang-db) [reply mlflow/metrics/base.py]: I've updated the documentation but since I could only get the documentation to work by putting it directly in the .rst file, the docstrings are duplicates and modifying them won't actually change the docs built - should I include a note here to modify documentation in the .rst file or remove the docstrings entirely?

File: mlflow/metrics/utils/metric_definitions.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+        _logger.warning(
+            f"Failed to load 'perplexity' metric (error: {e!r}), skipping metric logging."
+        )
+        return
+
+    _logger.info("Computing perplexity metric:")
+    results = perplexity.compute(predictions=predictions, model_id="gpt2")
+    return MetricValue(
+        scores=results["perplexities"], aggregate_results={"mean": results["mean_perplexity"]}
+    )
```
Comment (prithvikannan) [inline mlflow/metrics/utils/metric_definitions.py:72]: for a given set of `scores`, don't we know all the various aggregations (min, max, mean, median, variance, p90)? if so, does it make sense to compute and store all of the aggregations in the resulting `MetricValue`

this comment applies to the other metric eval_fns in this file too

File: mlflow/metrics/utils/metric_definitions.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+    scores = [textstat.automated_readability_index(prediction) for prediction in predictions]
+    return MetricValue(scores=scores, aggregate_results={"mean": sum(scores) / len(scores)})
+
+
+def _accuracy_eval_fn(eval_df, metrics):
+    if "target" in eval_df:
+        from sklearn.metrics import accuracy_score
+
+        acc = accuracy_score(y_true=eval_df["target"], y_pred=eval_df["prediction"])
+        return MetricValue(aggregate_results={"": acc})
```
Comment (prithvikannan) [inline mlflow/metrics/utils/metric_definitions.py:118]: is this supposed to be an empty string?

File: mlflow/metrics/utils/metric_definitions.py [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+
+
+def _rouge1_eval_fn(eval_df, metrics):
+    if "target" in eval_df:
```
Comment (prithvikannan) [inline mlflow/metrics/utils/metric_definitions.py:122]: what happens if `"target" not in eval_df`? should we throw some error or maybe emit some warning message

File: tests/metrics/test_metric_definitions.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        }
+    )
+    result = accuracy.eval_fn(eval_df, metrics={})
+    assert result.aggregate_results[""] == 0.5
+
+
+def test_rouge1():
+    eval_df = pd.DataFrame({"prediction": ["a", "d c"], "target": ["d", "b c"]})
+    result = rouge1.eval_fn(eval_df, metrics={})
+    assert result.aggregate_results["mean"] == 0.25
```
Comment (prithvikannan) [inline tests/metrics/test_metric_definitions.py:127]: should we also check the `result.values`?

File: mlflow/metrics/utils/metric_definitions.py (hunk 5 above)
Comment (annzhang-db) [reply mlflow/metrics/utils/metric_definitions.py:121]: right now (in mlflow.evaluate()) there is no error or warning message - we just don't calculate rouge if there is no target. i don't think we should throw an error because if there is no target the user shouldn't expect there to be a rouge score anyways - we could add a warning message

File: mlflow/metrics/utils/metric_definitions.py (hunk 3 above)
Comment (annzhang-db) [reply mlflow/metrics/utils/metric_definitions.py:71]: yeah right now i'm moving over all the current functionality from mlflow.evaluate() to use make_metric() without adding/changing anything that is calculated - planning to have a follow-up PR to calculate/store those aggregations

File: mlflow/metrics/utils/metric_definitions.py (hunk 4 above)
Comment (annzhang-db) [reply mlflow/metrics/utils/metric_definitions.py:117]: Yes, wasn't sure what to else to put here considering there are no scores and it is just an aggregation - I was thinking that metrics will be named metric_name_aggregation_name/version so an empty string would mean that it just takes the metric name.

File: mlflow/metrics/utils/metric_definitions.py (hunk 3 above)
Comment (prithvikannan) [reply mlflow/metrics/utils/metric_definitions.py:71]: ah understood. that makes sense.

File: mlflow/metrics/utils/metric_definitions.py (hunk 5 above)
Comment (prithvikannan) [reply mlflow/metrics/utils/metric_definitions.py:121]: i see, this makes sense.

**Review by prithvikannan — APPROVED** [review]
LGTM, thanks @annzhang-db
